File: packages/camera/camera_avfoundation/ios/Classes/CameraProperties.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "CameraProperties.h"

#pragma mark - flash mode

FLTFlashMode FLTGetFLTFlashModeForString(NSString *mode) {
  if ([mode isEqualToString:@"off"]) {
    return FLTFlashModeOff;
  } else if ([mode isEqualToString:@"auto"]) {
    return FLTFlashModeAuto;
  } else if ([mode isEqualToString:@"always"]) {
    return FLTFlashModeAlways;
  } else if ([mode isEqualToString:@"torch"]) {
    return FLTFlashModeTorch;
  } else {
    return FLTFlashModeInvalid;
  }
}

AVCaptureFlashMode FLTGetAVCaptureFlashModeForFLTFlashMode(FLTFlashMode mode) {
  switch (mode) {
    case FLTFlashModeOff:
      return AVCaptureFlashModeOff;
    case FLTFlashModeAuto:
      return AVCaptureFlashModeAuto;
    case FLTFlashModeAlways:
      return AVCaptureFlashModeOn;
    case FLTFlashModeTorch:
    default:
      return -1;
  }
}

#pragma mark - exposure mode

FCPPlatformExposureMode FCPGetExposureModeForString(NSString *mode) {
  if ([mode isEqualToString:@"auto"]) {
    return FCPPlatformExposureModeAuto;
  } else if ([mode isEqualToString:@"locked"]) {
    return FCPPlatformExposureModeLocked;
  } else {
    // This should be unreachable; see _serializeExposureMode in avfoundation_camera.dart.
    NSCAssert(false, @"Unsupported exposure mode");
    return FCPPlatformExposureModeAuto;
  }
}

#pragma mark - focus mode

FCPPlatformFocusMode FCPGetFocusModeForString(NSString *mode) {
  if ([mode isEqualToString:@"auto"]) {
    return FCPPlatformFocusModeAuto;
  } else if ([mode isEqualToString:@"locked"]) {
    return FCPPlatformFocusModeLocked;
  } else {
    // This should be unreachable; see _serializeFocusMode in avfoundation_camera.dart.
    NSCAssert(false, @"Unsupported focus mode");
    return FCPPlatformFocusModeAuto;
  }
}

#pragma mark - device orientation

UIDeviceOrientation FLTGetUIDeviceOrientationForString(NSString *orientation) {
  if ([orientation isEqualToString:@"portraitDown"]) {
    return UIDeviceOrientationPortraitUpsideDown;
  } else if ([orientation isEqualToString:@"landscapeLeft"]) {
    return UIDeviceOrientationLandscapeLeft;
  } else if ([orientation isEqualToString:@"landscapeRight"]) {
    return UIDeviceOrientationLandscapeRight;
  } else if ([orientation isEqualToString:@"portraitUp"]) {
    return UIDeviceOrientationPortrait;
  } else {
    return UIDeviceOrientationUnknown;
  }
}

FCPPlatformDeviceOrientation FCPGetPigeonDeviceOrientationForOrientation(
    UIDeviceOrientation orientation) {
  switch (orientation) {
    case UIDeviceOrientationPortraitUpsideDown:
      return FCPPlatformDeviceOrientationPortraitDown;
    case UIDeviceOrientationLandscapeLeft:
      return FCPPlatformDeviceOrientationLandscapeLeft;
    case UIDeviceOrientationLandscapeRight:
      return FCPPlatformDeviceOrientationLandscapeRight;
    case UIDeviceOrientationPortrait:
    default:
      return FCPPlatformDeviceOrientationPortraitUp;
  };
}

#pragma mark - video format

OSType FLTGetVideoFormatFromString(NSString *videoFormatString) {
  if ([videoFormatString isEqualToString:@"bgra8888"]) {
    return kCVPixelFormatType_32BGRA;
  } else if ([videoFormatString isEqualToString:@"yuv420"]) {
    return kCVPixelFormatType_420YpCbCr8BiPlanarVideoRange;
  } else {
    NSLog(@"The selected imageFormatGroup is not supported by iOS. Defaulting to brga8888");
    return kCVPixelFormatType_32BGRA;
  }
}

#pragma mark - file format

FCPFileFormat FCPGetFileFormatFromString(NSString *fileFormatString) {
  if ([fileFormatString isEqualToString:@"jpg"]) {
    return FCPFileFormatJPEG;
  } else if ([fileFormatString isEqualToString:@"heif"]) {
    return FCPFileFormatHEIF;
  } else {
    return FCPFileFormatInvalid;
  }
}

File: packages/camera/camera_avfoundation/darwin/camera_avfoundation/Sources/camera_avfoundation/CameraProperties.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import AVFoundation
import UIKit

/// Gets AVCaptureFlashMode from PlatformFlashMode.
/// mode - flash mode.
func getAVCaptureFlashMode(for mode: PlatformFlashMode) -> AVCaptureDevice.FlashMode {
  switch mode {
  case .off:
    return .off
  case .auto:
    return .auto
  case .always:
    return .on
  case .torch:
    assertionFailure("This mode cannot be converted, and requires custom handling.")
    return .off
  @unknown default:
    assertionFailure("Unknown flash mode")
    return .off
  }
}

/// Gets UIDeviceOrientation from its Pigeon representation.
/// orientation - the Pigeon device orientation.
func getUIDeviceOrientation(
  for orientation: PlatformDeviceOrientation
) -> UIDeviceOrientation {
  switch orientation {
  case .portraitDown:
    return .portraitUpsideDown
  case .landscapeLeft:
    return .landscapeLeft
  case .landscapeRight:
    return .landscapeRight
  case .portraitUp:
    return .portrait
  @unknown default:
    assertionFailure("Unknown device orientation")
    return .portrait
  }
}

/// Gets a Pigeon representation of UIDeviceOrientation.
/// orientation - the UIDeviceOrientation.
func getPigeonDeviceOrientation(
  for orientation: UIDeviceOrientation
) -> PlatformDeviceOrientation {
  switch orientation {
  case .portraitUpsideDown:
    return .portraitDown
  case .landscapeLeft:
    return .landscapeLeft
  case .landscapeRight:
    return .landscapeRight
  case .portrait:
    return .portraitUp
  default:
    return .portraitUp
  }
}

/// Gets pixel format from its Pigeon representation.
/// imageFormat - the Pigeon image format.
func getPixelFormat(for imageFormat: PlatformImageFormatGroup) -> OSType {
  switch imageFormat {
  case .bgra8888:
    return kCVPixelFormatType_32BGRA
  case .yuv420:
    return kCVPixelFormatType_420YpCbCr8BiPlanarVideoRange
  @unknown default:
    assertionFailure("Unknown image format")
    return kCVPixelFormatType_32BGRA
  }
}

/// Gets video stabilization mode from its Pigeon representation.
/// videoStabilizationMode - the Pigeon video stabilization mode.
func getAvCaptureVideoStabilizationMode(
  _ videoStabilizationMode: PlatformVideoStabilizationMode
) -> AVCaptureVideoStabilizationMode {

  switch videoStabilizationMode {
  case .off:
    return .off
  case .standard:
    return .standard
  case .cinematic:
    return .cinematic
  case .cinematicExtended:
    if #available(iOS 13.0, *) {
      return .cinematicExtended
    } else {
      return .cinematic
    }
  @unknown default:
    return .off
  }
}
